src/io/io_behaviorsaver.js

```diff
@@ -58,13 +58,89 @@ IO.BehaviorSaver = new (function() {
 		UI.Settings.updateBehaviorlib();
 		UI.Tools.notifyRosCommand('save');
-	}
+    }
+
+    var saveOnRobot = function() {
+        var names = Behavior.createNames();
+        var package_name = names.rosnode_name;
+        var ssh_username = UI.Settings.getRobotUsername();
+        var ssh_hostname = UI.Settings.getRobotHostname();
+        var ssh_package_path = ""
+
+        if(ssh_username === "" || ssh_hostname === "" /*|| ssh_password === ""*/) {
+            T.logError("SSH configurations are not well defined, fix them from Configuration Tab -> Code Generation Box");
+            return;
+        }
+
+        var ssh_password = window.prompt("Please enter {0}:{1} password: ".format(ssh_username, ssh_hostname));
+        if(ssh_password == null || ssh_password == "") {
+            return;
+        }
 
+        ROS.getPackagePath(package_name, (package_path) =>
+        {
+            try
+            {
+                var SSHClient = require('ssh2').Client;
+                conn = new SSHClient();
+                conn.on('ready', function() {
+                  var ros_distro = process.env.ROS_DISTRO;
+                  conn.exec('bash -c "source .profile && rospack find {1}"'.format(ros_distro, package_name), function(err, stream) {
+                    if (err) {
+                        T.logError(err);
+                    }
+                    stream.on('close', function(code, signal) {
+                      conn.end();
+                    })
+                    .on('data', function(remote_package_path) {
+                      ssh_package_path = String(remote_package_path).trim();
+                      T.logInfo('Path to package on the robot: ' + ssh_package_path);
+                      conn.end()
+                      var scp_client = require('scp2')
+                      var ssh_query = '{0}:{1}@{2}:{3}/'.format(
+                          ssh_username,
+                          ssh_password,
+                          ssh_hostname,
+                          ssh_package_path);
+                      scp_client.scp(package_path+"/", String(ssh_query), function(err) {
+                        var err_str = String(err);
+                        if( err_str != 'undefined') {
+                            T.logError("scp failed "+err);
+                        }
+                      });
+                    }).stderr.on('data', function(data) {
+                      T.logError(data+", Make sure robot's ROS environment is configured properly in \".profile\" script");
+                    })
+                  });
+                })
+                .on('error', function(err) {
+                        T.logError(err);
+                        T.logError("Please check your ssh configuration and entered password.");
+                 })
+                 .connect({
+                  host: ssh_hostname,
+                  username: ssh_username,
+                  password: ssh_password
+                });
+            }
```

Comment:
I'm not sure if I understand the above code block correctly. To me it looks like there is an ssh connection created first and then, within this connection, another scp connection is opened to transfer the file. Do I get it wrong? Or what is the reason for this approach?